### PR TITLE
Feature: Initial GitHub sync

### PR DESCRIPTION
**Pull request by chriku**

### CODE REVIEW

File: github/src/main/kotlin/gropius/Application.kt [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import org.springframework.boot.autoconfigure.SpringBootApplication
+import org.springframework.boot.context.event.ApplicationReadyEvent
+import org.springframework.boot.runApplication
+import org.springframework.context.annotation.Configuration
+import org.springframework.context.event.EventListener
+import org.springframework.data.mongodb.repository.config.EnableReactiveMongoRepositories
+import org.springframework.stereotype.Component
+
+@Configuration
+class ApplicationFisch(
```
Comment (nk-coding) [inline github/src/main/kotlin/gropius/Application.kt:15]: Do you want to keep this class?

File: github/src/main/kotlin/gropius/sync/github/Grabber.kt [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+import gropius.sync.github.generated.fragment.MetaData
+import gropius.sync.github.generated.fragment.PageInfoData
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.toList
+import org.bson.types.ObjectId
+import java.time.OffsetDateTime
+
+/**
+ * Requests data from github using steps that are managed over restarts for processing items (e.g. issues, timeline, ...)
```
Comment (nk-coding) [inline github/src/main/kotlin/gropius/sync/github/Grabber.kt:11]: missing documentation for generic type parameter

File: github/src/main/kotlin/gropius/sync/github/Grabber.kt [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+ */
+abstract class Grabber<T : Any> {
+
+    /**
+     * The response of a single step
+     */
+    interface StepResponse<T : Any> {
```
Comment (nk-coding) [inline github/src/main/kotlin/gropius/sync/github/Grabber.kt:18]: not sure, but doesn't this generic parameter shadow the one from the outer class?

File: github/src/main/kotlin/gropius/sync/github/IssueGrabber.kt [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+    /**
+     * test repository user
+     */
+    val user = "espressif"
+
+    /**
+     * test repository name
+     */
+    val repo = "idf-component-manager" //"llvm-project"
```
Comment (nk-coding) [inline github/src/main/kotlin/gropius/sync/github/IssueGrabber.kt:40]: what about this inline comment?

File: github/src/main/kotlin/gropius/sync/github/IssueGrabber.kt [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+    class IssueStepResponse(val content: IssueReadQuery.Data) : StepResponse<IssueDataExtensive> {
```
Comment (nk-coding) [inline github/src/main/kotlin/gropius/sync/github/IssueGrabber.kt:42]: missing documentation

File: github/src/main/kotlin/gropius/sync/github/MongoConfig.kt [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+import org.springframework.data.mongodb.core.convert.MongoCustomConversions
+import java.time.OffsetDateTime
+import java.time.ZoneOffset
+import java.util.*
+
+/**
+ * Custom configuration for the mongodb driver implementing custom conversion for offsetdatetime
+ */
+@Configuration
+class MongoConfig {
```
Comment (nk-coding) [inline github/src/main/kotlin/gropius/sync/github/MongoConfig.kt:15]: missing documentation for inner functions / classes

File: github/src/main/kotlin/gropius/sync/github/TimelineGrabber.kt [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+import org.springframework.data.mongodb.core.*
+import org.springframework.data.mongodb.core.query.Criteria
+import org.springframework.data.mongodb.core.query.Query
+import org.springframework.data.mongodb.core.query.Update
+import java.time.OffsetDateTime
+
+/**
+ * Implementation of Grabber to retrieve timeline items and cache them in the database
+ */
+class TimelineGrabber(
```
Comment (nk-coding) [inline github/src/main/kotlin/gropius/sync/github/TimelineGrabber.kt:26]: missing documentation for params, inner class

File: github/src/main/kotlin/gropius/sync/github/model/TimelineEventInfo.kt [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+import org.springframework.data.annotation.Id
+import org.springframework.data.mongodb.core.index.Indexed
+import org.springframework.data.mongodb.core.mapping.Document
+import java.time.OffsetDateTime
+
+/**
+ * Mapping of a single timeline event from neo4j to github
+ */
+@Document
+data class TimelineEventInfo(
```
Comment (nk-coding) [inline github/src/main/kotlin/gropius/sync/github/model/TimelineEventInfo.kt:13]: missing documentation for properties

File: github/src/main/kotlin/gropius/sync/github/repository/IssueInfoRepository.kt [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+/**
+ * Mapping of a single issue from neo4j to github
+ */
+@Repository
+interface IssueInfoRepository : ReactiveMongoRepository<IssueInfo, ObjectId> {
+    /**
+     * Find all issues that have outstanding changes
+     */
+    suspend fun findByDirtyIsTrue(): Flow<IssueInfo>
```
Comment (nk-coding) [inline github/src/main/kotlin/gropius/sync/github/repository/IssueInfoRepository.kt:17]: missing documentation for params, return

File: sync/src/main/kotlin/gropius/sync/IssueCleaner.kt [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+ *
+ * Sync services using this can only zip the timelines together and clean it into a valid state to merge two issues
+ */
+@Component
+class IssueCleaner(
+    @Qualifier("graphglueNeo4jOperations")
+    val neoOperations: ReactiveNeo4jOperations,
+) {
+
+    private suspend fun cleanLabels(issue: Issue) {
```
Comment (nk-coding) [inline sync/src/main/kotlin/gropius/sync/IssueCleaner.kt:25]: missing documentation

**Review by nk-coding — CHANGES_REQUESTED** [review]

File: github/src/main/kotlin/gropius/sync/github/Grabber.kt (hunk 3 above)
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/Grabber.kt:18]: the generic of the class is not available in the interface

File: github/src/main/kotlin/gropius/sync/github/IssueGrabber.kt (hunk 4 above)
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/IssueGrabber.kt:40]: an alternative test repo. I use this to switch between many issues and few issues until you finished your mutations

File: github/src/main/kotlin/gropius/sync/github/IssueGrabber.kt (hunk 5 above)
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/IssueGrabber.kt:52]: cafe999d546dae9937ec2e811c2d873877c50a22

File: github/src/main/kotlin/gropius/sync/github/TimelineGrabber.kt (hunk 7 above)
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/TimelineGrabber.kt:26]: cafe999d546dae9937ec2e811c2d873877c50a22

File: github/src/main/kotlin/gropius/sync/github/model/TimelineEventInfo.kt (hunk 8 above)
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/model/TimelineEventInfo.kt:13]: cafe999d546dae9937ec2e811c2d873877c50a22

File: sync/src/main/kotlin/gropius/sync/IssueCleaner.kt (hunk 10 above)
Comment (chriku) [reply sync/src/main/kotlin/gropius/sync/IssueCleaner.kt:25]: `private`

File: github/src/main/kotlin/gropius/sync/github/repository/IssueInfoRepository.kt (hunk 9 above)
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/repository/IssueInfoRepository.kt:18]: 4428e298d423fb8732889ff48146dcffe4724df9

File: github/src/main/kotlin/gropius/sync/github/MongoConfig.kt (hunk 6 above)
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/MongoConfig.kt:16]: f7b2dc1b32aedbf244d06473b330a1c761273056 If you have an idea of the content, please feel welcome to add docs

File: github/src/main/kotlin/gropius/sync/github/Grabber.kt (hunk 3 above)
Comment (nk-coding) [reply github/src/main/kotlin/gropius/sync/github/Grabber.kt]: a possible solution would be to make the interface an inner abstract class
but I don't think that's worth

File: sync/src/main/kotlin/gropius/sync/IssueCleaner.kt (hunk 10 above)
Comment (nk-coding) [reply sync/src/main/kotlin/gropius/sync/IssueCleaner.kt:25]: please still document

File: github/src/main/kotlin/gropius/sync/github/TimelineGrabber.kt (hunk 7 above)
Comment (nk-coding) [reply github/src/main/kotlin/gropius/sync/github/TimelineGrabber.kt:26]: the parameter if the inner class is still not documented

File: sync/src/main/kotlin/gropius/sync/IssueCleaner.kt (hunk 10 above)
Comment (chriku) [reply sync/src/main/kotlin/gropius/sync/IssueCleaner.kt:29]: ab9c57cd07bf8db9b681e5d747a90adb87a2d1b3

File: github/src/main/kotlin/gropius/sync/github/repository/IssueInfoRepository.kt [hunk 11]
```diff
@@ -0,0 +1,35 @@
+package gropius.sync.github.repository
+
+import gropius.sync.github.model.IssueInfo
+import kotlinx.coroutines.flow.Flow
+import org.bson.types.ObjectId
+import org.springframework.data.mongodb.repository.ReactiveMongoRepository
+import org.springframework.stereotype.Repository
+
+/**
+ * Mapping of a single issue from neo4j to github
```
Comment (ModellbahnFreak) [inline github/src/main/kotlin/gropius/sync/github/repository/IssueInfoRepository.kt:10]: Maybe something along the lines of "repository for retrieving the mappings..."

This comment implies this is the model class.

File: github/src/main/kotlin/gropius/sync/github/repository/LabelInfoRepository.kt [hunk 12]
```diff
@@ -0,0 +1,26 @@
+package gropius.sync.github.repository
+
+import gropius.sync.github.model.LabelInfo
+import org.bson.types.ObjectId
+import org.springframework.data.mongodb.repository.ReactiveMongoRepository
+import org.springframework.stereotype.Repository
+
+/**
+ * Mapping of a single label from neo4j to github
```
Comment (ModellbahnFreak) [inline github/src/main/kotlin/gropius/sync/github/repository/LabelInfoRepository.kt:9]: Maybe something along the lines of "repository for retrieving the mappings..."

This comment implies this is the model class.

File: github/src/main/kotlin/gropius/sync/github/repository/RepositoryInfoRepository.kt [hunk 13]
```diff
@@ -0,0 +1,20 @@
+package gropius.sync.github.repository
+
+import gropius.sync.github.model.RepositoryInfo
+import org.bson.types.ObjectId
+import org.springframework.data.mongodb.repository.ReactiveMongoRepository
+import org.springframework.stereotype.Repository
+
+/**
+ * Mapping of a single repository from neo4j to github
```
Comment (ModellbahnFreak) [inline github/src/main/kotlin/gropius/sync/github/repository/RepositoryInfoRepository.kt:9]: Maybe something along the lines of "repository for retrieving the mappings..."

This comment implies this is the model class.

File: github/src/main/kotlin/gropius/sync/github/repository/TimelineEventInfoRepository.kt [hunk 14]
```diff
@@ -0,0 +1,25 @@
+package gropius.sync.github.repository
+
+import gropius.sync.github.model.TimelineEventInfo
+import org.bson.types.ObjectId
+import org.springframework.data.mongodb.repository.ReactiveMongoRepository
+import org.springframework.stereotype.Repository
+
+/**
+ * Mapping of a single timeline event from neo4j to github
```
Comment (ModellbahnFreak) [inline github/src/main/kotlin/gropius/sync/github/repository/TimelineEventInfoRepository.kt:9]: Maybe something along the lines of "repository for retrieving the mappings..."

This comment implies this is the model class.

File: github/src/main/kotlin/gropius/sync/github/repository/UserInfoRepository.kt [hunk 15]
```diff
@@ -0,0 +1,25 @@
+package gropius.sync.github.repository
+
+import gropius.sync.github.model.UserInfo
+import org.bson.types.ObjectId
+import org.springframework.data.mongodb.repository.ReactiveMongoRepository
+import org.springframework.stereotype.Repository
+
+/**
+ * Mapping of a single user from neo4j to github
```
Comment (ModellbahnFreak) [inline github/src/main/kotlin/gropius/sync/github/repository/UserInfoRepository.kt:9]: Maybe something along the lines of "repository for retrieving the mappings..."

This comment implies this is the model class.

**Review by ModellbahnFreak — APPROVED** [review]
I am really sorry that it took me so long to review this.
I just found some little things in the doc comments, thats why I'm approving it already.

I didn't mean to completly ignore this...

File: github/src/main/kotlin/gropius/sync/github/Grabber.kt [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+     * Remove a single id from the database cache
+     * @param node The id in the node to query (not mongo id)
+     */
+    protected abstract suspend fun removeFromCache(node: String)
+
+    /**
+     * Increase the number of failed attempts for a node in the cache
+     * @param node Node id to increase the number on (not mongo id)
+     */
+    protected abstract suspend fun increasedFailedCache(node: String)
```
[the remark this answers is not in the document]
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/Grabber.kt]: 7d0512574a5e62a269f543a1bdf377f95a905bbf

File: github/src/main/kotlin/gropius/sync/github/repository/IssueInfoRepository.kt (hunk 11 above)
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/repository/IssueInfoRepository.kt]: 93208eeec463b1a5f6cd2de7a2af126618451613

File: github/src/main/kotlin/gropius/sync/github/repository/LabelInfoRepository.kt (hunk 12 above)
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/repository/LabelInfoRepository.kt]: 93208eeec463b1a5f6cd2de7a2af126618451613

File: github/src/main/kotlin/gropius/sync/github/repository/RepositoryInfoRepository.kt (hunk 13 above)
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/repository/RepositoryInfoRepository.kt]: 93208eeec463b1a5f6cd2de7a2af126618451613

File: github/src/main/kotlin/gropius/sync/github/repository/TimelineEventInfoRepository.kt (hunk 14 above)
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/repository/TimelineEventInfoRepository.kt]: 93208eeec463b1a5f6cd2de7a2af126618451613

File: github/src/main/kotlin/gropius/sync/github/repository/UserInfoRepository.kt (hunk 15 above)
Comment (chriku) [reply github/src/main/kotlin/gropius/sync/github/repository/UserInfoRepository.kt]: 93208eeec463b1a5f6cd2de7a2af126618451613

**Comment by ModellbahnFreak** [comment]
All good from me

**Pull request merged**